src/commands/crashes/upload-symbols.ts

```diff
@@ -0,0 +1,420 @@
+        return this.readZipFileToBuffer(path);
+      default:
+        // File doesn't points to correct symbols
+        throw new UploadSymbolsError(ErrorCodes.InvalidParameter, `${path} is not a valid symbols file/directory`);
+    }
+  }
+
+  private async executeSymbolsUploadingBeginRequest(client: MobileCenterClient, app: DefaultApp): Promise<models.SymbolUploadBeginResponse>{
+    debug("Executing API request to get uploading URL");
+    let symbolUploadingBeginRequest = clientRequest<models.SymbolUploadBeginResponse>(cb => client.crashOperations.postSymbolUpload(
+      app.ownerName, 
+      app.appName, 
+      "Apple", 
+      cb)).catch((error: any) => {
+          return Promise.reject(`failed to start the symbol uploading: ${this.getErrnoExceptionString(error)}`);
+      });
+
+    debug("Analyzing upload start request response status code");
+    const uploadingBeginResponse = await symbolUploadingBeginRequest;
+    const uploadingBeginStatusCode = uploadingBeginResponse.response.statusCode;
+    const uploadingBeginStatusMessage = uploadingBeginResponse.response.statusMessage;
+    if (uploadingBeginStatusCode >= 400) {
+      throw new UploadSymbolsError(ErrorCodes.Exception, 
+        `the symbol upload begin API request was rejected: HTTP ${uploadingBeginStatusCode} - ${uploadingBeginStatusMessage}`);
+    }
+
+    return uploadingBeginResponse.result;
+  }
+
+  private uploadZipFile(uploadUrl: string, zippedFileBuffer: Buffer, md5Hash: string): Promise<void>{
+    debug("Uploading the prepared ZIP file");
+    return new Promise<void>((resolve, reject) => {
+      Request.put(uploadUrl, {
+        body: zippedFileBuffer,
+        headers: {
+          "Content-Length": zippedFileBuffer.length,
+          "Content-MD5": md5Hash,
+          "x-ms-blob-type": "BlockBlob",
+          "Content-Type": "application/zip"
+        }
+      })
+      .on("error", (error) => {
+        reject(new UploadSymbolsError(ErrorCodes.Exception, `Zip file uploading failed: ${error.message}`));
+      })
+      .on("response", (response) => {
+        if(response.statusCode < 400){
+          resolve()
+        }
+        else{
+          reject(new UploadSymbolsError(ErrorCodes.Exception, `Zip file uploading failed: HTTP ${response.statusCode}`));
+        }
+      });
+    });
+  }
+
+  private async executeSymbolsUploadingEndRequest(client: MobileCenterClient, app: DefaultApp, symbolUploadId: string, desiredStatus: SymbolsUploadEndRequestStatus): Promise<models.SymbolUpload>{
+    debug(`Finishing symbols uploading with desired status: ${desiredStatus}`);
+    let symbolUploadingEndRequest = clientRequest<models.SymbolUpload>(cb => client.crashOperations.patchSymbolUpload(
+      symbolUploadId,
+      app.ownerName,
+      app.appName,
+      desiredStatus,
+      cb, 
+    )).catch((error: any) => {
+        return Promise.reject(`failed to finalize the symbol upload with status '${desiredStatus}': ${this.getErrnoExceptionString(error)}`);
```

Comment:
Throw an exception here instead of calling Promise.reject. That way handling the failure will consistently be an Error object.